File: lambda/lambda_function.py
import os
import json
import openai
import time

openai.api_key = os.environ["OPENAI_API_KEY"]
ASSISTANT_ID = os.environ["OPENAI_ASSISTANT_ID"]

def lambda_handler(event, context):
    if event.get("httpMethod", "") == "OPTIONS":
        return {
            "statusCode": 200,
            "headers": {
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Content-Type",
                "Access-Control-Allow-Methods": "OPTIONS,POST"
            },
            "body": ""
        }
    # Parse messages from the request body
    body = json.loads(event["body"])
    messages = body.get("messages", [])

    # Create a new thread for each chat session (stateless example)
    thread = openai.beta.threads.create()
    thread_id = thread.id

    # Add all user and assistant messages to the thread
    for msg in messages:
        if msg["role"] in ["user", "assistant"]:
            openai.beta.threads.messages.create(
                thread_id=thread_id,
                role=msg["role"],
                content=msg["content"]
            )

    # Run the assistant on the thread
    run = openai.beta.threads.runs.create(
        thread_id=thread_id,
        assistant_id=ASSISTANT_ID
    )

    # Wait for the run to complete (polling)
    while True:
        run_status = openai.beta.threads.runs.retrieve(thread_id=thread_id, run_id=run.id)
        if run_status.status in ["completed", "failed", "cancelled"]:
            break
        time.sleep(1)

    # Get the latest assistant message
    thread_messages = openai.beta.threads.messages.list(thread_id=thread_id)
    assistant_message = next(
        (m for m in reversed(thread_messages.data) if m.role == "assistant"), None
    )

    content = assistant_message.content[0].text.value if assistant_message else "Sorry, I couldn't generate a response."

    return {
        "statusCode": 200,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",  # Or your domain instead of *
            "Access-Control-Allow-Headers": "Content-Type",
            "Access-Control-Allow-Methods": "OPTIONS,POST"
        },
        "body": json.dumps({"content": content})
    }

File: lambda/lambda_function.js
const OpenAI = require('openai');
const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const corsHeaders = {
  "Content-Type": "application/json",
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "OPTIONS,POST",
  "Access-Control-Allow-Headers": "Content-Type,x-api-key,x-thread-id"
};

exports.handler = async (event) => {
  try {
    const requestMethod = event.httpMethod || event.requestContext?.http?.method;

    if (requestMethod === 'OPTIONS') {
      return {
        statusCode: 204,
        headers: corsHeaders,
        body: ''
      };
    }

    let body = event.body;
    if (typeof body === 'string') {
      body = body ? JSON.parse(body) : {};
    } else if (!body || typeof body !== 'object') {
      body = {};
    }

    const input = body.input; // or body.messages, depending on your frontend

    if (!input || typeof input !== 'string') {
      return {
        statusCode: 400,
        headers: corsHeaders,
        body: JSON.stringify({
          content: 'Invalid request payload.',
          response: 'Invalid request payload.'
        })
      };
    }

    // Create a new thread
    const thread = await openai.beta.threads.create();

    // Add the user message to the thread
    await openai.beta.threads.messages.create(thread.id, {
      role: 'user',
      content: input
    });

    // Run the assistant and poll for completion
    const run = await openai.beta.threads.runs.createAndPoll(
      thread.id,
      { assistant_id: process.env.OPENAI_ASSISTANT_ID }
    );

    // Get the latest assistant message
    let cleanedContent = "Sorry, I couldn't generate a response.";
    if (run.status === 'completed') {
      const messagesResponse = await openai.beta.threads.messages.list(thread.id);
      const assistantMessage = messagesResponse.data
        .slice()
        .reverse()
        .find(m => m.role === "assistant");

      if (assistantMessage) {
        cleanedContent = assistantMessage.content[0].text.value.replace(/【[^】]*】/g, '').trim();
      }
    }

    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({ content: cleanedContent, response: cleanedContent, id: thread.id })
    };
  } catch (err) {
    console.error('Error communicating with AI service:', err);
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        content: "Sorry, I couldn't reach the AI service.",
        response: "Sorry, I couldn't reach the AI service."
      })
    };
  }
};